src/main/java/com/redhat/devtools/lsp4ij/features/semanticTokens/viewProvider/LSPSemanticTokensFileViewProviderHelper.java

```diff
@@ -189,19 +191,62 @@ LSPSemanticToken getSemanticToken(int offset) {
         // If this file has semantic tokens, use them
         Map<Integer, LSPSemanticToken> semanticTokensByOffset = getSemanticTokensByOffset();
         if (!ContainerUtil.isEmpty(semanticTokensByOffset)) {
-            return semanticTokensByOffset.get(offset);
+            LSPSemanticToken semanticToken = semanticTokensByOffset.get(offset);
+            // Update the view provider's effective offset as appropriate
+            setEffectiveOffset(semanticToken == null ? offset : -1);
+            return semanticToken;
         }
         // Otherwise stub a semantic token for the entire file so that it won't highlight as a link on mouse hover
         else {
             // By caching the stub on the file this way, it's automatically evicted when the file changes
-            return CachedValuesManager.getCachedValue(file, new CachedValueProvider<>() {
+            LSPSemanticToken fileLevelSemanticToken = CachedValuesManager.getCachedValue(file, new CachedValueProvider<>() {
                 @Override
                 @NotNull
                 public Result<LSPSemanticToken> compute() {
                     LSPSemanticToken stubSemanticToken = new LSPSemanticToken(file, file.getTextRange(), null, null);
                     return Result.create(stubSemanticToken, file);
                 }
             });
+            // Update the file-level token's requested offset
+            fileLevelSemanticToken.setLastRequestedOffset(offset);
+            return fileLevelSemanticToken;
         }
     }
+
+    /**
+     * Stores the effective offset as a thread local.
+     *
+     * @param offset the effective offset
+     */
+    private void setEffectiveOffset(int offset) {
+        effectiveOffsetPtr.set(offset);
+    }
+
+    @Override
+    public int getEffectiveOffset(@NotNull PsiElement element) {
```

Comment:
And this is how we compute it. If it's for a real element, return the effective offset of that element. If it's not, return the last requested offset from the view provider. If neither, just return -1 denoting that there's no effective offset.